### ISSUE TITLE
Add wrapper or header when image is zoomed in

### ISSUE THREAD

**Issue by PaoloDiBello** [rpearce/react-medium-image-zoom#332]
- Feature request
## Description
It would be nice if we could have a header when the image is zoomed in, not to confuse  with the caption which is when the image is zoomed out.
#242 

I would be really nice if we could have a new prop such as ZoomWrapperComponent 
that lets you wrap the zoomed in images 

```
<Zoom
...
ZoomWrapperComponent={children => {
<div>{children}</div>
}} >
```
or at least lets you add a component on top of the image

```
<Zoom
ZoomHeaderComponent={HeaderComponent}>
```

Thanks

**Comment by rpearce** [comment]
@PaoloDiBello This is an interesting idea! I definitely want to have _some_ way to provide a custom caption when an image is zoomed, and this idea feels like it is a step in the right direction.

I will think about it and try some things out.

**Comment by rpearce** [comment]
@PaoloDiBello Update: I'm playing with something to allow designs like this (the design would be up to the user):

<img width="1277" alt="imagen" src="https://user-images.githubusercontent.com/592876/183259665-7f22e912-55a3-46bb-a198-0b7f536d2670.png">

**Comment by rpearce** [comment]
@PaoloDiBello How does this API usage feel? This would allow you to reorder the image, the button, and put any other content in there that you might want. You could put content in-between them, before the image, after both of them, etc.

```tsx
export const CustomModalContent: ComponentStory<typeof Zoom> = (props) => (
  <div>
    <h1>Custom Modal Content</h1>
    <div className="mw-600">
      <Zoom {...props} ZoomContent={CustomZoomContent}>
        <img
          alt={imgThatWanakaTree.alt}
          src={imgThatWanakaTree.src}
          height="320"
          loading="lazy"
        />
      </Zoom>
    </div>
  </div>
)

const CustomZoomContent = ({ img, buttonUnzoom }) => {
  return <>
    <figure>
      {img}

      <figcaption className="my-figcaption">
        <p>
          That Wanaka Tree, also known as the Wanaka Willow, is a willow tree
          located at the southern end of Lake Wānaka in the Otago region of New
          Zealand.
        </p>
        <a href="https://en.wikipedia.org/wiki/That_Wanaka_Tree" className="my-figcaption-link">
          Wikipedia, That Wanaka Tree
        </a>
      </figcaption>
    </figure>

    {buttonUnzoom}
  </>
}
```

**Comment by PaoloDiBello** [comment]
I love the idea, it would be great to have something like this

**Comment by rpearce** [comment]
@PaoloDiBello Okay, great. I already have that working, but I'm
going to try to get feedback from some other people before moving forward.

**Comment by rpearce** [comment]
This is my branch I'm working on, if you want to see any details: https://github.com/rpearce/react-medium-image-zoom/compare/feature/custom-modal-content?expand=1

**Comment by rpearce** [comment]
@PaoloDiBello @tommoor Here's the code and screen recording of my current example. This wraps the image in a `<figure>` and provides a user-defined `<figcaption>` that has different positions based on the image's dimensional ratio vs the browser's dimensional ratio at different screen widths. That's just an example, though; you could do anything you wanted.

If you like it, I could mark this `ZoomContent` property as "experimental" and do a beta release for y'all to test it out.

What do you think?

<details>
  <summary>Expand to see code</summary>

```tsx
export const CustomModalContent: ComponentStory<typeof Zoom> = (props) => (
  <div>
    <h1>Custom Modal Content</h1>
    <div className="mw-600">
      <Zoom {...props} ZoomContent={CustomZoomContent}>
        <img
          alt={imgThatWanakaTree.alt}
          src={imgThatWanakaTree.src}
          height="320"
          loading="lazy"
        />
      </Zoom>
    </div>
  </div>
)

// NOTE: onUnzoom isn't used here, but you could use this
// with your own button as a way to close the zoom modal.
// Thanks to @tommoor for the suggestion.
const CustomZoomContent = ({ buttonUnzoom, img, onUnzoom }) => {
  const { width: imgWidth, height: imgHeight } = img?.props ?? {}

  const classCaption = useMemo(() => {
    let cn = 'zoom-caption'

    if (imgWidth && imgHeight) {
      cn = cn + ' ' + (
        imgWidth / imgHeight > window.innerWidth / window.innerHeight
          ? 'zoom-caption--bottom'
          : 'zoom-caption--left'
      )
    }

    return cn
  }, [imgWidth, imgHeight])

  return <>
    {buttonUnzoom}

    <figure>
      {img}
      <figcaption className={classCaption}>
        That Wanaka Tree, also known as the Wanaka Willow, is a willow tree
        located at the southern end of Lake Wānaka in the Otago region of New
        Zealand.
        <cite className="zoom-caption-cite">
          Wikipedia, <a className="zoom-caption-link" href="https://en.wikipedia.org/wiki/That_Wanaka_Tree">
            That Wanaka Tree
          </a>
        </cite>
      </figcaption>
    </figure>
  </>
}
```

</details>

https://user-images.githubusercontent.com/592876/183328045-10ee3b45-2e14-480e-8698-d9f6f4624e58.mp4

**Comment by tommoor** [comment]
👍 how does the animation feel at what point does the caption appear? Do you think it could make sense to gradually fade the caption in through the zoom animation?

**Comment by rpearce** [comment]
> 👍 how does the animation feel at what point does the caption appear? Do you think it could make sense to gradually fade the caption in through the zoom animation?

Since that is completely defined by you, I think you could make it do whatever you like in your ZoomContent component. All of the transitions are done through overridable CSS, so it'd be up to you.

**Comment by rpearce** [comment]
There are a couple of more things to take into account, and I'm too tired to finish it tonight. I'll try to get something up tomorrow night or the next night! It's feeling alright so far.

**Comment by rpearce** [comment]
@tommoor I think you're going to like this...

https://user-images.githubusercontent.com/592876/184269505-463e4dae-5d09-4311-a5a7-ee20131a7d0b.mp4

I'm sending through the internal `modalState` to the `ZoomComponent` and using it like this

```tsx
useLayoutEffect(() => {
  if (modalState === 'LOADED') {
    setIsLoaded(true)
  } else if (modalState === 'UNLOADING') {
    setIsLoaded(false)
  }
}, [modalState])
```

to set different classes based on state.

**Comment by rpearce** [comment]
Has anyone had a chance to play with that beta version yet?

**Comment by tommoor** [comment]
Following along but I haven't had time to jump back on this project yet 😅

**Comment by rpearce** [comment]
Updated this to `react-medium-image-zoom@5.1.0-beta.1`, for it now includes the `gatsby-plugin-image` fix from `v5.0.2`.

**Comment by rpearce** [comment]
Update: I didn't know I needed a break, but I did.

I'm going to try to fix the `zoomMargin` bug mentioned in https://github.com/rpearce/react-medium-image-zoom/issues/350 this weekend — we'll see how that works out?

Here's my plan:
* Fix `zoomMargin`, ship a patch version
* Get this code merged and another thing merged, then ship a minor version

**Comment by rpearce** [comment]
@PaoloDiBello @tommoor This has been released in `v5.1.0`!

I added y'all to the All Contributors section in the README for your contributions. Thank you for your time 💜
